fiftyone/core/video.py

```diff
@@ -0,0 +1,582 @@
+            self._source_collection._dataset.add_frame_field(
+                field_name, **field_kwargs
+            )
+
+        if delete:
+            for field_name in delete_fields:
+                self._source_collection._dataset.delete_frame_field(field_name)
+
+
+def make_frames_dataset(
+    sample_collection,
+    sample_frames=True,
+    frames_patt=None,
+    size=None,
+    min_size=None,
+    max_size=None,
+    force_sample=False,
+    name=None,
+):
+    """Creates a dataset that contains one sample per video frame in the
+    collection.
+
+    The returned dataset will contain all frame-level fields and the ``tags``
+    of each video as sample-level fields, as well as ``sample_id`` and
+    ``frame_id`` fields that record the IDs of the frame in the input
+    collection.
+
+    When ``sample_frames`` is True (the default), this method samples each
+    video in the collection into a directory of per-frame images with the same
+    basename as the input video with frame numbers/format specified by
+    ``frames_patt``.
+
+    For example, if ``frames_patt = "%%06d.jpg"``, then videos with the
+    following paths::
+
+        /path/to/video1.mp4
+        /path/to/video2.mp4
+        ...
+
+    would be sampled as follows::
+
+        /path/to/video1/
+            000001.jpg
+            000002.jpg
+            ...
+        /path/to/video2/
+            000001.jpg
+            000002.jpg
+            ...
+
+    .. note::
+
+        The returned dataset is independent from the source collection;
+        modifying it will not affect the source collection.
+
+    Args:
+        sample_collection: a
+            :class:`fiftyone.core.collections.SampleCollection`
+        sample_frames (True): whether to sample the video frames. If False,
+            the dataset cannot currently be viewed in the App
+        frames_patt (None): a pattern specifying the filename/format to use to
+            store the sampled frames, e.g., ``"%%06d.jpg"``. The default value
+            is ``fiftyone.config.default_sequence_idx + fiftyone.config.default_image_ext``
+        size (None): an optional ``(width, height)`` for each frame. One
+            dimension can be -1, in which case the aspect ratio is preserved
+        min_size (None): an optional minimum ``(width, height)`` for each
+            frame. A dimension can be -1 if no constraint should be applied.
+            The frames are resized (aspect-preserving) if necessary to meet
+            this constraint
+        max_size (None): an optional maximum ``(width, height)`` for each
+            frame. A dimension can be -1 if no constraint should be applied.
+            The frames are resized (aspect-preserving) if necessary to meet
+            this constraint
+        force_sample (False): whether to resample videos whose sampled frames
+            already exist
+        name (None): a name for the returned dataset
```

Comment:
I think it would be nice to have a sparse option, where frames are only created for existing frame documents on the root dataset. And, perhaps, new frames can be propagated back to the root.

Noting it now, but perhaps this is a feature request for discussion.